crates/kornia-imgproc/src/features/fast.rs

```diff
@@ -1,162 +1,332 @@
-        cols - 3,      // 12
-        -cols - 3,     // 13
-        -2 * cols - 2, // 14
-        -3 * cols - 1, // 15
-        -3 * cols,     // 16
-    ];
-
-    // Process rows in parallel
-    let keypoints = (3..rows - 3)
-        .into_par_iter()
-        .flat_map(|y| {
-            let row_start_idx = y * cols;
-            let mut row_keypoints = Vec::new();
-
-            for x in 3..cols - 3 {
-                if is_fast_corner(
-                    src.as_slice(),
-                    row_start_idx + x,
-                    offsets,
-                    threshold,
-                    arc_length,
-                ) {
-                    row_keypoints.push([x, y]);
+// Using u8 constants can be slightly faster in tight loops than an enum.
+const PIXEL_SIMILAR: u8 = 0;
+const PIXEL_BRIGHTER: u8 = 1;
+const PIXEL_DARKER: u8 = 2;
+
+/// A FAST (Features from Accelerated Segment Test) feature detector for corner detection in images.
+#[derive(Clone)]
+pub struct FastDetector {
+    /// The intensity threshold for detecting corners.
+    pub threshold: f32,
+    /// The minimum distance between detected keypoints.
+    pub min_distance: usize,
+    /// The minimum arc length for a sequence of contiguous pixels to be considered a corner.
+    pub arc_length: usize,
+    corner_response: Image<f32, 1, CpuAllocator>,
+    taken: Vec<bool>,
+}
+
+impl FastDetector {
+    /// Creates a new `FastDetector` with the specified parameters.
+    ///
+    /// # Arguments
+    ///
+    /// * `image_size` - The size of the image to process.
+    /// * `threshold` - The intensity threshold for detecting corners.
+    /// * `arc_length` - The minimum arc length for a sequence of contiguous pixels to be considered a corner.
+    /// * `min_distance` - The minimum distance between detected keypoints.
+    ///
+    /// # Returns
+    ///
+    /// Returns a `Result` containing the new `FastDetector` or an `ImageError`.
+    pub fn new(
+        image_size: ImageSize,
+        threshold: f32,
+        arc_length: usize,
+        min_distance: usize,
+    ) -> Result<Self, ImageError> {
+        Ok(Self {
+            threshold,
+            min_distance,
+            arc_length,
+            corner_response: Image::from_size_val(image_size, 0.0, CpuAllocator)?,
+            taken: vec![false; image_size.height * image_size.width],
+        })
+    }
+
+    /// Clears the internal state of the detector, marking it ready to detect again.
+    pub fn clear(&mut self) {
+        self.taken.iter_mut().for_each(|px| *px = false);
+    }
+
+    /// Computes the corner response for the input image.
+    ///
+    /// # Arguments
+    ///
+    /// * `src` - The source grayscale image.
+    ///
+    /// # Returns
+    ///
+    /// Returns a reference to the image containing the corner response.
+    pub fn compute_corner_response<A: ImageAllocator>(
+        &mut self,
+        src: &Image<f32, 1, A>,
+    ) -> Result<&Image<f32, 1, CpuAllocator>, ImageError> {
```

Comment:
why do you need to return the image ? other parts of the pipeline can just access that. If the user for some reason need it, we can provide a getter method